### ISSUE TITLE
Cropped + Highlighted = wrong word highlighted

### ISSUE THREAD

**Issue by alexisvisco** [meilisearch/MeiliSearch#1283]
**Describe the bug**
When a text needs to be cropped and contain a highlighted element, the highlighted element is wrong (shifted).
 
**To Reproduce**
- create a index
- add a document with an attribute and a text with a lot of characters
- search through this index, add the attribute to be highlighted, and cropped.  

If the text need to be cropped, the highlighted result will be wrong. 

**Expected behavior**
Just the good word(s) should be highlighted. 

**Screenshots**
![image](https://user-images.githubusercontent.com/17145294/110356757-de3e1080-803a-11eb-8c24-942f05de7e62.png)
As you can see the query is "attendrir" but the highlighted word is "t agrém"

**Meilisearch version**
 - Engine 0.19

**Comment by curquiza** [comment]
Hello @alexisvisco thanks for your issue!
Can you send us the exact document so that we can copy/paste and use it for our tests! We are indeed re-doing the core of the search engine and this kind of issue should be fixed thanks to this!

**Comment by alexisvisco** [comment]
```json
{
   "hits":[
      {
         "id":"7166f71b-2d11-4628-afb7-6d16175273f2",
         "published_at":0,
         "title":"CHAR DE LA TURCHE 10m",
         "content":"Recette de sortie inoubliable pour 7 personnes (Gilles G - François - Isabelle - Pierre - Gerard - Gilles M - Vincent) Ingrédients :&nbsp; 7 personnes bien mûres !, Un glissement de terrain, Une neige béton &amp; ravinées par les pluies, Quelques  arbres au milieu du chemin, Un zeste de recherche d’itinéraire, Un rebrousse chemin Mise en œuvre : Faire miroiter une sortie sans imprévus aux&nbsp; personnes bien mures, agrémenter d’un glissement de terrain fermant la route &amp; rebrousser le chemin qui n’en était pas un !Saupoudrer le zeste d’itinéraire sur une forêt difficilement praticable pour l’attendrir et agrémenter le tout d’une neige béton et ravinée. Un fois le sommet bien cuit par un soleil de plomb et prêt à découper les pentes, penser à mettre un casque pour éviter les sapins qui traversent la table ! Bonne dégustation de ce Char de la Turche maison.",
         "owner_display_name":"alexis",
         "categories":[
            "escalade"
         ],
         "event_at_display_date":"samedi 6 mars",
         "categories_ids":[
            "cc32f3e8-4877-47b1-8f5c-ee410b3647c3"
         ],
         "owner_id":"5bc87cec-72e7-4921-bf85-770c13363a03",
         "published":"false",
         "thumbnail":"http://asvf-montagne.s3.fr-par.scw.cloud/asvf-montagne/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855_thumbnail_01_2.jpg",
         "event_at":1615040859,
         "created_at":1615203211,
         "updated_at":1615231071,
         "_formatted":{
            "id":"7166f71b-2d11-4628-afb7-6d16175273f2",
            "published_at":0,
            "title":"CHAR DE LA TURCHE 10m",
            "content":"d’un glissement de terrain fermant la route &amp; rebrousser le chemin qui n’en était pas un !Saupoudrer le zeste d’itinéraire sur une forêt difficilement praticable pour l’attendrir e<em>t agréme</em>nter le tout d’une neige béton et ravinée. Un fois le sommet bien cuit par un soleil de plomb et prêt à découper les pentes, penser à mettre un casque pour éviter les sapins qui traversent la table ! Bonne",
            "owner_display_name":"alexis",
            "categories":[
               "escalade"
            ],
            "event_at_display_date":"samedi 6 mars",
            "categories_ids":[
               "cc32f3e8-4877-47b1-8f5c-ee410b3647c3"
            ],
            "owner_id":"5bc87cec-72e7-4921-bf85-770c13363a03",
            "published":"false",
            "thumbnail":"http://asvf-montagne.s3.fr-par.scw.cloud/asvf-montagne/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855_thumbnail_01_2.jpg",
            "event_at":1615040859,
            "created_at":1615203211,
            "updated_at":1615231071
         }
      }
   ],
   "offset":0,
   "limit":20,
   "nbHits":1,
   "exhaustiveNbHits":false,
   "processingTimeMs":46,
   "query":"attendrir"
}
```

**Comment by alexisvisco** [comment]
@curquiza I think it's related to accents.

**Comment by curquiza** [comment]
Thanks @alexisvisco!

**Comment by timoschwarzer** [comment]
Seems to be a duplicate of https://github.com/meilisearch/MeiliSearch/issues/1173

**Comment by CaroFG** [comment]
I have tested it with a text in English, no special characters and I encounter the same problem...

**Comment by curquiza** [comment]
Could you detail your steps @CaroFG?

**Comment by CaroFG** [comment]
I've taken the movies dataset, I added an extra paragraph to the overview of the movie 'Shazam', to make it long enough. So now, the movie looks like this: 

`    {
      "id": 287947,
      "title": "Shazam",
      "poster": "https://image.tmdb.org/t/p/w1280/xnopI5Xtky18MPhK40cZAGAOVeV.jpg",
      "overview": "A Paul Atreides, a brilliant and gifted young man born into a great destiny beyond his understanding, must travel to the most dangerous planet in the universe to ensure the future of his family and his people. As malevolent forces explode into conflict over the planets exclusive supply of the most precious resource in existence—a commodity capable of unlocking humanitys greatest potential—only those who can conquer their fear will survive.  boy is given the ability to become an adult superhero in times of need with a single magic word.",
      "release_date": "2019-03-23"
     }
`

Then I run the following: 

`~ curl -X POST 'http://localhost:7700/indexes/movies/search' \
  --data '{ "q": "adult superhero", "attributesToHighlight": ["*"], "attributesToCrop": ["overview"], "cropLength": 10 }`

And the formatted result looks like this:

`      "_formatted": {
        "id": 287947,
        "title": "Shazam",
        "poster": "https://image.tmdb.org/t/p/w1280/xnopI5Xtky18MPhK40cZAGAOVeV.jpg",
        "overview": "an adul<em>t sup</em>erhero",
        "release_date": "2019-03-23"
      }`

As you can see, instead of having `<em> adult super hero </em>`, I got: `<em>t sup</em>`

**Comment by robsontenorio** [comment]
@curquiza any chances to be included on v0.20? Just saw some RC tags today :)

**Comment by curquiza** [comment]
Hello @robsontenorio!
Sorry, it will be in the v0.21.0 but not in the v0.20.0!
FYI, the v0.21.0 will be a new search engine that would fix many issues (including a lot of highlight issues) you can see in this board -> https://github.com/meilisearch/MeiliSearch/projects/2

**Comment by timoschwarzer** [comment]
Whoa! Didn't expect the new engine _that_ fast, you people rock! Thanks for your work! :rocket:

**Comment by robsontenorio** [comment]
@curquiza is there any build available for 0.21.0 ? Or are you focusing exclusively in 0.20.0 development ?

**Comment by curquiza** [comment]
@robsontenorio we are focusing on the v0.21.0 implementation. For the moment the crop and the highlight are not finished so I cannot share any v0.21.0 binary that fixes this issue at the moment.
